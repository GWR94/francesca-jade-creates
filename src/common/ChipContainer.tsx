import React from "react";
import { Chip, ThemeProvider, createMuiTheme, makeStyles } from "@material-ui/core";

interface ChipContainerProps {
  tags: string[];
  type: "Cake" | "Creates";
}

// create theme to style the tags
const chipTheme = createMuiTheme({
  palette: {
    primary: {
      main: "#9370f6",
    },
    secondary: {
      main: "#ff80f7",
    },
  },
});

// create styles for component
const useStyles = makeStyles({
  container: {
    display: "inline-flex",
    justifyContent: "center",
    alignItems: "center",
    flexWrap: "wrap",
    width: "100%",
  },
});

/**
 * A component which follows material-ui's outlined input where you can add tags
 * to an individual product.
 * @param {string[]} tags - Array of string containing the labels for each tag
 * @param {"Cake" | "Creates"} type - The type of product which the tags are for.
 */
const ChipContainer: React.SFC<ChipContainerProps> = ({ tags, type }): JSX.Element => {
  // use styles
  const classes = useStyles();
  return (
    <div className={classes.container}>
      {/* The theme changes the tags colour based on which type the product is */}
      <ThemeProvider theme={chipTheme}>
        {/* map the tags array into each individual chip component */}
        {tags.map(
          (tag, i): JSX.Element => (
            <Chip
              key={i}
              className="tags__tag"
              label={tag}
              size="small"
              color={type === "Cake" ? "secondary" : "primary"}
              style={{ color: "#fff", borderRadius: "3px", alignItems: "center" }}
            />
          ),
        )}
      </ThemeProvider>
    </div>
  );
};

export default ChipContainer;
