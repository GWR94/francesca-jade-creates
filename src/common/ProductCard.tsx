import React, { useState } from "react";
import { Card, Button, Alert, Tag } from "@blueprintjs/core";
import { API, graphqlOperation } from "aws-amplify";
import { S3Image } from "aws-amplify-react";
import { useHistory } from "react-router-dom";
import { ProductCardProps } from "./interfaces/Product.i";
import { deleteProduct } from "../graphql/mutations";
import { Toaster } from "../utils/index";

const Product: React.FC<ProductCardProps> = ({ product, admin }): JSX.Element => {
  const history = useHistory();
  const { id, image, title, price, shippingCost, type, tags } = product;
  const [deleteAlertOpen, setDeleteAlert] = useState(false);

  const handleDeleteProduct = async (): Promise<void> => {
    try {
      await API.graphql(graphqlOperation(deleteProduct, { input: { id } }));
      setDeleteAlert(false);
      Toaster.show({
        intent: "success",
        message: `${title} has been successfully removed.`,
      });
    } catch (err) {
      console.error(err);
      Toaster.show({
        intent: "danger",
        message: `${title} could not be removed.
        Please try again`,
      });
    }
  };

  const getIcon = (type): JSX.Element => {
    switch (type) {
      case "Cake":
        return <i className="fas fa-birthday-cake" style={{ color: "#fd4ef2" }} />;
      case "Card":
        return <i className="fas fa-pencil-alt" style={{ color: "#9370f6" }} />;
      case "Frame":
        return <i className="fas fa-palette" style={{ color: "#69abec" }} />;
      default:
        return null;
    }
  };

  return (
    <>
      <Card elevation={2} interactive className="product__card">
        <div className="product__text-container">
          <p>
            <strong>{title}</strong> - {getIcon(type)}
          </p>
          <p>
            {price > 0
              ? `£${price.toFixed(2)} + £${shippingCost.toFixed(2)} postage`
              : "No set price"}
          </p>
          {tags && (
            <div>
              {tags.map(
                (tag, i): JSX.Element => (
                  <Tag active key={i} style={{ margin: "0px 4px 4px" }}>
                    {tag}
                  </Tag>
                ),
              )}
            </div>
          )}
        </div>
        <div className="product__image-container">
          <S3Image
            imgKey={image[0]?.key}
            theme={{
              photoImg: {
                width: "100%",
                display: "flex",
                alignItems: "center",
                justifyContent: "center",
              },
            }}
          />
        </div>
        <div className="new-product__button-container">
          {!admin ? (
            <Button
              intent="success"
              text={price === 0 ? "Request a quote" : "Pay with Stripe"}
              icon="credit-card"
              style={{ margin: "8px 4px 0" }}
            />
          ) : (
            <>
              <Button
                text="Delete"
                intent="danger"
                onClick={(): void => setDeleteAlert(true)}
                style={{ margin: "8px 4px 0" }}
              />
              <Button
                text="Edit"
                intent="warning"
                onClick={(): void => history.push(`/account/${product.id}`)}
                style={{ margin: "8px 4px 0" }}
              />
            </>
          )}
        </div>
      </Card>
      <Alert
        isOpen={deleteAlertOpen}
        cancelButtonText="Cancel"
        confirmButtonText="Delete"
        icon="trash"
        intent="danger"
        onCancel={(): void => setDeleteAlert(false)}
        onConfirm={handleDeleteProduct}
      >
        <p>Are you sure you want to delete &quot;{title}&quot;?</p>
        <p>This cannot be undone.</p>
      </Alert>
    </>
  );
};

export default Product;
