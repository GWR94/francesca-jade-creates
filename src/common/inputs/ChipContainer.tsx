import React from "react";
import { Chip, ThemeProvider, makeStyles } from "@material-ui/core";
import { COLORS, breakpoints } from "../../themes";

interface ChipContainerProps {
  tags: string[];
  type: "Cake" | "Creates";
}

/**
 * TODO
 * [ ] Create chip input component and remove mui-chip-input library
 */

/**
 * A component which follows material-ui's outlined input where you can add tags
 * to an individual product.
 * @param tags - Array of string containing the labels for each tag
 * @param type - The type of product which the tags are for.
 */
const ChipContainer: React.FC<ChipContainerProps> = ({ tags, type }): JSX.Element => {
  // create styles for component
  const useStyles = makeStyles({
    container: {
      display: "inline-flex",
      justifyContent: "center",
      alignItems: "center",
      flexWrap: "wrap",
      width: "100%",
      marginTop: 10,
    },
    tag: {
      cursor: "pointer",
      color: "#fff",
      borderRadius: 3,
      padding: "8px 12px",
      fontSize: "1rem",
      margin: "4px",
      [breakpoints.down("xs")]: {
        padding: "4px 8px",
        fontSize: "0.8rem",
      },
    },
    createsTag: {
      background: COLORS.DarkGrey,
      "&:hover": {
        background: COLORS.Gray,
      },
    },
    cakesTag: {
      background: COLORS.DarkPink,
      "&:hover": {
        background: COLORS.Pink,
      },
    },
  });
  // use styles
  const classes = useStyles();
  return (
    <div className={classes.container}>
      {/* The theme changes the tags colour based on which type the product is */}
      {/* map the tags array into each individual chip component */}
      {tags.map(
        (tag, i): JSX.Element => (
          <Chip
            key={i}
            className={`${classes.tag} ${
              type === "Cake" ? classes.cakesTag : classes.createsTag
            }`}
            label={tag}
            size="small"
            color={type === "Cake" ? "secondary" : "primary"}
            style={{
              color: "#fff",
              borderRadius: "3px",
              alignItems: "center",
            }}
          />
        ),
      )}
    </div>
  );
};

export default ChipContainer;
