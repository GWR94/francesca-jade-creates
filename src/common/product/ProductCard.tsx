import React, { useState } from "react";
import { API, graphqlOperation } from "aws-amplify";
import {
  Card,
  Button,
  Dialog,
  DialogTitle,
  DialogActions,
  DialogContent,
} from "@material-ui/core";
import { S3Image } from "aws-amplify-react";
import { ProductCardProps } from "../interfaces/Product.i";
import { deleteProduct } from "../../graphql/mutations";
import { Toaster } from "../../utils/index";
import TagsInput from "../TagsInput";

const Product: React.FC<ProductCardProps> = ({
  product,
  admin = false,
  history,
}): JSX.Element => {
  const { id, image, title, price, shippingCost, type, tags } = product;
  const [deleteAlertOpen, setDeleteAlert] = useState(false);

  const handleDeleteProduct = async (): Promise<void> => {
    try {
      await API.graphql(graphqlOperation(deleteProduct, { input: { id } }));
      setDeleteAlert(false);
      Toaster.show({
        intent: "success",
        message: `${title} has been successfully removed.`,
      });
    } catch (err) {
      console.error(err);
      Toaster.show({
        intent: "danger",
        message: `${title} could not be removed.
        Please try again`,
      });
    }
  };

  const getIcon = (type): JSX.Element => {
    switch (type) {
      case "Cake":
        return (
          <i
            className="fas fa-birthday-cake product__icon"
            style={{ color: "#fd4ef2" }}
            tabIndex={0}
            role="button"
            onClick={(e): void => {
              e.stopPropagation();
              history.push("/cakes");
            }}
          />
        );
      case "Creates":
        return (
          <i
            className="fas fa-pencil-alt product__icon"
            style={{ color: "#9370f6" }}
            tabIndex={0}
            role="button"
            onClick={(e): void => {
              e.stopPropagation();
              history.push("/creates");
            }}
          />
        );
      default:
        return null;
    }
  };

  return (
    <>
      <Card
        elevation={2}
        style={{ padding: "20px" }}
        className="product__card animate fadeIn"
        onClick={(): void =>
          history.push(`/${type === "Cake" ? "cakes" : "creates"}/${id}`)
        }
      >
        {getIcon(type)}
        <div className="product__text-container">
          <div className="product__title">
            <strong>{title}</strong>
          </div>
          <p className="product__price">
            {price > 0
              ? `£${price.toFixed(2)} + £${shippingCost.toFixed(2)} postage`
              : "Customer requests quote"}
          </p>
          {tags && <TagsInput type={type} tags={tags} />}
        </div>
        <div className="product__image-container">
          <S3Image
            imgKey={image[0]?.key}
            theme={{
              photoImg: {
                width: "100%",
                display: "flex",
                alignItems: "center",
                justifyContent: "center",
              },
            }}
          />
        </div>
        {admin && (
          <div className="new-product__button-container">
            <Button
              onClick={(e): void => {
                e.stopPropagation();
                setDeleteAlert(true);
              }}
              style={{ margin: "8px 4px 0", background: "#fd4ef2", color: "#fff" }}
            >
              Delete
            </Button>
            <Button
              onClick={(e): void => {
                e.stopPropagation();
                history.push(`/account/${product.id}`);
              }}
              style={{ margin: "8px 4px 0", background: "#ff80f7", color: "#fff" }}
            >
              Edit
            </Button>
          </div>
        )}
      </Card>
      <Dialog open={deleteAlertOpen} onClose={(): void => setDeleteAlert(false)}>
        <DialogTitle>Delete &quot;{title}&quot;?</DialogTitle>
        <DialogContent>
          <p>Are you sure you want to delete &quot;{title}&quot;?</p>
          <p>This cannot be undone.</p>
        </DialogContent>
        <DialogActions>
          <Button color="primary" onClick={(): void => setDeleteAlert(false)}>
            Cancel
          </Button>
          <Button color="secondary" onClick={handleDeleteProduct}>
            Confirm
          </Button>
        </DialogActions>
      </Dialog>
    </>
  );
};

export default Product;
