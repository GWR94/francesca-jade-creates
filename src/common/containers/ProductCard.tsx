import React, { useState } from "react";
import { API, graphqlOperation } from "aws-amplify";
import {
  Card,
  Button,
  Dialog,
  DialogTitle,
  DialogActions,
  DialogContent,
  CardHeader,
  IconButton,
  makeStyles,
  CardMedia,
  CardContent,
  MenuItem,
  Menu,
  Fab,
  Tooltip,
} from "@material-ui/core";
import { useDispatch } from "react-redux";
import { Skeleton } from "@material-ui/lab";
import { MoreVert, AddShoppingCartOutlined } from "@material-ui/icons";
import { S3Image } from "aws-amplify-react";
import { useHistory } from "react-router-dom";
import { openSnackbar } from "../../utils/Notifier";
import * as actions from "../../actions/basket.actions";
import { ProductCardProps } from "../../pages/accounts/interfaces/Product.i";
import { deleteProduct } from "../../graphql/mutations";
import ChipContainer from "../inputs/ChipContainer";
import { INTENT } from "../../themes";
import styles from "../styles/productCard.style";

const Product: React.FC<ProductCardProps> = ({ product, admin = false }): JSX.Element => {
  // create styles for component
  const useStyles = makeStyles({
    ...styles,
    fab: {
      position: "absolute",
      bottom: 4,
      right: 4,
      backgroundColor:
        product.type === "Cake" ? "rgba(253, 78, 242, 0.4)" : "rgba(147, 112, 246, 0.5)",
      "&:hover": {
        backgroundColor:
          product.type === "Cake"
            ? "rgba(253, 78, 242, 0.65)"
            : "rgba(147, 112, 246, 0.75)",
      },
    },
  });

  const classes = useStyles();
  // destructure product for ease of variable use
  const { id, images, title, type, tags, tagline, variants } = product;
  // boolean which shows/hides delete alert visibility
  const [deleteAlertOpen, setDeleteAlert] = useState(false);
  // boolean which opens dropdown menu for admin options
  const [menuOpen, setMenuOpen] = useState(false);
  // boolean which shows/hides loading UI effects
  const [isLoading, setLoading] = useState(true);
  // import and initialise useHistory() for navigation around the page
  const history = useHistory();
  // create anchorRef to allow a point to fit the anchor point for the menu
  const anchorRef = React.useRef<SVGSVGElement>(null);
  // connect with redux via hook.
  const dispatch = useDispatch();

  /**
   * Function to delete the current product from the database, using the deleteProduct
   * graphQL mutation.
   */
  const handleDeleteProduct = async (): Promise<void> => {
    try {
      // use the id of the current product as the input for deleteProduct
      await API.graphql(
        graphqlOperation(deleteProduct, {
          input: {
            id,
          },
        }),
      );
      // close the confirm delete dialog
      setDeleteAlert(false);
      // notify the user of success using a success snackbar with a relevant title.
      openSnackbar({
        severity: INTENT.Success,
        message: `${title} has been successfully removed.`,
      });
    } catch (err) {
      console.error(err);
      // if there are any errors, notify the user with a danger snackbar.
      openSnackbar({
        severity: INTENT.Danger,
        message: `${title} could not be removed.
        Please try again`,
      });
    }
  };

  const handleGetPrices = (): string => {
    let min = Infinity;
    if (variants?.length) {
      for (const variant of variants) {
        min = Math.min(min, variant.price.item);
      }
    }
    if (min === Infinity) return `Variable Price - Request a Quote!`;
    return `From Â£${min.toFixed(2)}`;
  };

  return (
    <>
      <Card
        elevation={2}
        onClick={(): void =>
          history.push(`/${type === "Cake" ? "cakes" : "creates"}/${id}`)
        }
        className={classes.card}
      >
        <CardHeader
          classes={{
            title: classes.headerContainer,
          }}
          action={
            // if loading return a skeleton of the potential product
            admin && (
              <IconButton
                aria-label="extra options"
                onClick={(e): void => {
                  e.stopPropagation();
                  setMenuOpen(true);
                }}
                className={classes.options}
              >
                <MoreVert ref={anchorRef} />
              </IconButton>
            )
          }
          title={
            // if loading return a skeleton of the potential product
            isLoading ? (
              <Skeleton
                animation="wave"
                style={{
                  marginRight: 14,
                }}
              />
            ) : (
              title
            )
          }
          subheader={isLoading ? <Skeleton animation="wave" /> : tagline || ""}
          style={{
            textAlign: "center",
          }}
        />
        <CardContent
          classes={{
            root: classes.root,
          }}
        >
          {isLoading ? (
            <div
              style={{
                margin: "30px 0",
              }}
            >
              <Skeleton
                animation="wave"
                height={10}
                style={{
                  marginBottom: 6,
                }}
              />
              <Skeleton
                animation="wave"
                height={10}
                width="80%"
                style={{
                  margin: "auto",
                }}
              />
            </div>
          ) : (
            <>
              <p className={classes.price}>{handleGetPrices()}</p>
              {tags && (
                <div
                  style={{
                    marginBottom: 10,
                  }}
                >
                  <ChipContainer type={type} tags={tags} />
                </div>
              )}
            </>
          )}
          <CardMedia className={classes.media} title={title}>
            <S3Image
              imgKey={images.collection[0]?.key}
              theme={{
                photoImg: isLoading
                  ? {
                      display: "none",
                    }
                  : {
                      width: "100%",
                      display: "flex",
                      alignItems: "center",
                      justifyContent: "center",
                    },
              }}
              onLoad={(): void => setLoading(false)}
            />
            {/* if loading return a skeleton of the potential product */}
            {isLoading && (
              <Skeleton
                animation="wave"
                variant="rect"
                style={{
                  width: "100%",
                  height: 400,
                }}
              />
            )}
          </CardMedia>
          <Tooltip title="Add to Shopping Basket" arrow placement="top">
            <Fab
              aria-label="Add to shopping basket"
              className={classes.fab}
              onClick={(e): void => {
                e.stopPropagation();
                try {
                  dispatch(
                    actions.addToBasket({
                      ...product,
                      image: images.collection[images.cover],
                    }),
                  );
                  openSnackbar({
                    message: `Added ${product.title} to basket.`,
                    severity: "success",
                  });
                } catch (err) {
                  openSnackbar({
                    message: `Unable to add ${product.title} to basket. Please try again.`,
                    severity: "error",
                  });
                }
              }}
            >
              <AddShoppingCartOutlined />
            </Fab>
          </Tooltip>
        </CardContent>
      </Card>
      {admin && (
        <>
          <Menu
            open={menuOpen}
            anchorEl={anchorRef.current}
            onClose={(): void => setMenuOpen(false)}
            transformOrigin={{
              vertical: -32,
              horizontal: -20,
            }}
          >
            <MenuItem
              onClick={(e): void => {
                e.stopPropagation();
                history.push(`/account/${product.id}`);
              }}
            >
              Edit Product
            </MenuItem>
            <MenuItem
              onClick={(e): void => {
                e.stopPropagation();
                setMenuOpen(false);
                setDeleteAlert(true);
              }}
            >
              Delete Product
            </MenuItem>
          </Menu>
          <Dialog open={deleteAlertOpen} onClose={(): void => setDeleteAlert(false)}>
            <DialogTitle>
              Delete &quot;
              {title}
              &quot;?
            </DialogTitle>
            <DialogContent>
              <p>
                Are you sure you want to delete &quot;
                {title}
                &quot;?
              </p>
              <p>This cannot be undone.</p>
            </DialogContent>
            <DialogActions>
              <Button color="primary" onClick={(): void => setDeleteAlert(false)}>
                Cancel
              </Button>
              <Button color="secondary" onClick={handleDeleteProduct}>
                Confirm
              </Button>
            </DialogActions>
          </Dialog>
        </>
      )}
    </>
  );
};

export default Product;
