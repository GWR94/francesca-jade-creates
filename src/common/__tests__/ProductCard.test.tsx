import React, { useState as useStateMock } from "react";
import { shallow } from "enzyme";
import { Card, Button } from "@material-ui/core";
import { S3Image } from "aws-amplify-react";
import ProductCard from "../containers/ProductCard";
import { history } from "../../routes/Router";
import ChipContainer from "../inputs/ChipContainer";
import { cakeProduct, createsProduct } from "../../__mocks__/products";

jest.mock("react", () => ({
  ...jest.requireActual("react"),
  useState: jest.fn(),
}));

describe("<ProductCard />", () => {
  const setState = jest.fn();

  beforeEach(() => {
    // @ts-ignore
    useStateMock.mockImplementation((init) => [init, setState]);
  });

  it('should render correctly when user IS admin and the product is of type "Cake"', () => {
    // create shallow wrapper
    const wrapper = shallow(<ProductCard product={cakeProduct} admin />);
    // match snapshot
    expect(wrapper).toMatchSnapshot();
    // find card
    const card = wrapper.find(Card);
    // expect one instance of card to be in wrapper.
    expect(wrapper.find(Card).length).toBe(1);
    // expect the birthday cake icon when the product type is a cake
    expect(card.childAt(0).prop("className")).toBe("fas fa-birthday-cake product__icon");
    // expect the correct icon styles for a cake
    expect(card.childAt(0).prop("style")).toStrictEqual({
      color: "#fd4ef2",
    });
    // expect the correct title
    expect(wrapper.find(".product__title").text()).toBe("Racing Cake");
    // expect the correct price
    expect(wrapper.find(".product__price").text()).toBe("Customer requests quote");
    expect(wrapper.find(S3Image).length).toBe(1);
    expect(wrapper.find(S3Image).prop("imgKey")).toBe(cakeProduct.image[0].key);
    // expect the tags to be correct for the current product
    expect(wrapper.find(ChipContainer).prop("tags")).toStrictEqual(["Cake", "Racing"]);
    // expect there to be two children in the button container
    expect(wrapper.find(".new-product__button-container").children().length).toBe(2);
    // expect both children to be of type Button.
    expect(wrapper.find(".new-product__button-container").childAt(0).type()).toBe(Button);
    expect(wrapper.find(".new-product__button-container").childAt(1).type()).toBe(Button);
    // expect first button to have "Delete" text, which is an admin only option
    expect(wrapper.find(Button).at(0).prop("text")).toBe("Delete");
    // expect first button to have "Edit" text, which is an admin only option
    expect(wrapper.find(Button).at(1).prop("text")).toBe("Edit");
    // mock stopPropagation so it doesn't fail in test
    const stopPropagation = jest.fn();
    // simulate delete button click
    wrapper.find(Button).at(0).simulate("click", {
      stopPropagation,
    });
    // expect setState (alertOpen) to have been called one time with true as the parameter.
    expect(setState).toHaveBeenCalledWith(true);
  });

  it('should render correctly when user IS admin and the product is of type "Creates"', () => {
    // create shallow wrapper
    const wrapper = shallow(
      <ProductCard product={createsProduct} admin history={history} />,
    );
    // match snapshot
    expect(wrapper).toMatchSnapshot();
    // find card
    const card = wrapper.find(Card);
    // expect one instance of card to be in wrapper.
    expect(wrapper.find(Card).length).toBe(1);
    // expect the pencil icon when the product type is a creation.
    expect(card.childAt(0).prop("className")).toBe("fas fa-pencil-alt product__icon");
    // expect the correct icon styles for a cake
    expect(card.childAt(0).prop("style")).toStrictEqual({
      color: "#9370f6",
    });
    // expect the correct title
    expect(wrapper.find(".product__title").text()).toBe("Name Frame");
    // expect the correct price
    expect(wrapper.find(".product__price").text()).toBe("£14.50 + £4.50 postage");
    expect(wrapper.find(S3Image).length).toBe(1);
    expect(wrapper.find(S3Image).prop("imgKey")).toBe(createsProduct.image[0].key);
    // expect the tags to be correct for the current product
    expect(wrapper.find(ChipContainer).prop("tags")).toStrictEqual([
      "Personalised",
      "Name",
      "Frame",
    ]);
    // expect there to be two children in the button container
    expect(wrapper.find(".new-product__button-container").children().length).toBe(2);
    // expect both children to be of type Button.
    expect(wrapper.find(".new-product__button-container").childAt(0).type()).toBe(Button);
    expect(wrapper.find(".new-product__button-container").childAt(1).type()).toBe(Button);
    // expect first button to have "Delete" text, which is an admin only option
    expect(wrapper.find(Button).at(0).prop("text")).toBe("Delete");
    // expect first button to have "Edit" text, which is an admin only option
    expect(wrapper.find(Button).at(1).prop("text")).toBe("Edit");
    // mock stopPropagation so it doesn't fail in test
    const stopPropagation = jest.fn();
    // simulate delete button click
    wrapper.find(Button).at(0).simulate("click", {
      stopPropagation,
    });
    // expect setState (alertOpen) to have been called one time with true as the parameter.
    expect(setState).toHaveBeenCalledWith(true);
  });

  it("should render correctly when the user is NOT an admin and the product is of type 'Cake'", () => {
    // create shallow wrapper
    const wrapper = shallow(<ProductCard product={cakeProduct} history={history} />);
    // match snapshot
    expect(wrapper).toMatchSnapshot();
    // find card
    const card = wrapper.find(Card);
    // expect one instance of card to be in wrapper.
    expect(wrapper.find(Card).length).toBe(1);
    // expect the birthday cake icon when the product type is a cake
    expect(card.childAt(0).prop("className")).toBe("fas fa-birthday-cake product__icon");
    // expect the correct icon styles for a cake
    expect(card.childAt(0).prop("style")).toStrictEqual({
      color: "#fd4ef2",
    });
    // expect the correct title
    expect(wrapper.find(".product__title").text()).toBe("Racing Cake");
    // expect the correct price
    expect(wrapper.find(".product__price").text()).toBe("Customer requests quote");
    expect(wrapper.find(S3Image).length).toBe(1);
    expect(wrapper.find(S3Image).prop("imgKey")).toBe(cakeProduct.image[0].key);
    // expect the tags to be correct for the current product
    expect(wrapper.find(ChipContainer).prop("tags")).toStrictEqual(["Cake", "Racing"]);
  });

  it("should render correctly when the user is NOT admin and product type is 'Creates'", () => {
    const wrapper = shallow(<ProductCard product={createsProduct} history={history} />);
    // match snapshot
    expect(wrapper).toMatchSnapshot();
    // find card
    const card = wrapper.find(Card);
    // expect one instance of card to be in wrapper.
    expect(wrapper.find(Card).length).toBe(1);
    // expect the pencil icon when the product type is a creation.
    expect(card.childAt(0).prop("className")).toBe("fas fa-pencil-alt product__icon");
    // expect the correct icon styles for a cake
    expect(card.childAt(0).prop("style")).toStrictEqual({
      color: "#9370f6",
    });
    // expect the correct title
    expect(wrapper.find(".product__title").text()).toBe("Name Frame");
    // expect the correct price
    expect(wrapper.find(".product__price").text()).toBe("£14.50 + £4.50 postage");
    expect(wrapper.find(S3Image).length).toBe(1);
    expect(wrapper.find(S3Image).prop("imgKey")).toBe(createsProduct.image[0].key);
    // expect the tags to be correct for the current product
    expect(wrapper.find(ChipContainer).prop("tags")).toStrictEqual([
      "Personalised",
      "Name",
      "Frame",
    ]);
  });
});
