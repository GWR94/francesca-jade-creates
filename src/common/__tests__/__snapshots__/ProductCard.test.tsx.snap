// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> should render correctly when the user is NOT admin and product type is 'Creates' 1`] = `
<Fragment>
  <Blueprint3.Card
    className="product__card animate fadeIn"
    elevation={2}
    interactive={true}
    onClick={[Function]}
  >
    <i
      className="fas fa-pencil-alt product__icon"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "color": "#9370f6",
        }
      }
      tabIndex={0}
    />
    <div
      className="product__text-container"
    >
      <div
        className="product__title"
      >
        <strong>
          Name Frame
        </strong>
      </div>
      <p
        className="product__price"
      >
        £14.50 + £4.50 postage
      </p>
      <TagsInput
        tags={
          Array [
            "Personalised",
            "Name",
            "Frame",
          ]
        }
        type="Creates"
      />
    </div>
    <div
      className="product__image-container"
    >
      <S3Image
        imgKey="/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG"
        theme={
          Object {
            "photoImg": Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
              "width": "100%",
            },
          }
        }
      />
    </div>
  </Blueprint3.Card>
  <Blueprint3.Alert
    canEscapeKeyCancel={false}
    canOutsideClickCancel={false}
    cancelButtonText="Cancel"
    confirmButtonText="Delete"
    icon="trash"
    intent="danger"
    isOpen={false}
    onCancel={[Function]}
    onConfirm={[Function]}
  >
    <p>
      Are you sure you want to delete "
      Name Frame
      "?
    </p>
    <p>
      This cannot be undone.
    </p>
  </Blueprint3.Alert>
</Fragment>
`;

exports[`<ProductCard /> should render correctly when the user is NOT an admin and the product is of type 'Cake' 1`] = `
<Fragment>
  <Blueprint3.Card
    className="product__card animate fadeIn"
    elevation={2}
    interactive={true}
    onClick={[Function]}
  >
    <i
      className="fas fa-birthday-cake product__icon"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "color": "#fd4ef2",
        }
      }
      tabIndex={0}
    />
    <div
      className="product__text-container"
    >
      <div
        className="product__title"
      >
        <strong>
          Racing Cake
        </strong>
      </div>
      <p
        className="product__price"
      >
        Customer requests quote
      </p>
      <TagsInput
        tags={
          Array [
            "Cake",
            "Racing",
          ]
        }
        type="Cake"
      />
    </div>
    <div
      className="product__image-container"
    >
      <S3Image
        imgKey="/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357819265-racecarcrop.jpg"
        theme={
          Object {
            "photoImg": Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
              "width": "100%",
            },
          }
        }
      />
    </div>
  </Blueprint3.Card>
  <Blueprint3.Alert
    canEscapeKeyCancel={false}
    canOutsideClickCancel={false}
    cancelButtonText="Cancel"
    confirmButtonText="Delete"
    icon="trash"
    intent="danger"
    isOpen={false}
    onCancel={[Function]}
    onConfirm={[Function]}
  >
    <p>
      Are you sure you want to delete "
      Racing Cake
      "?
    </p>
    <p>
      This cannot be undone.
    </p>
  </Blueprint3.Alert>
</Fragment>
`;

exports[`<ProductCard /> should render correctly when user IS admin and the product is of type "Cake" 1`] = `
<Fragment>
  <Blueprint3.Card
    className="product__card animate fadeIn"
    elevation={2}
    interactive={true}
    onClick={[Function]}
  >
    <i
      className="fas fa-birthday-cake product__icon"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "color": "#fd4ef2",
        }
      }
      tabIndex={0}
    />
    <div
      className="product__text-container"
    >
      <div
        className="product__title"
      >
        <strong>
          Racing Cake
        </strong>
      </div>
      <p
        className="product__price"
      >
        Customer requests quote
      </p>
      <TagsInput
        tags={
          Array [
            "Cake",
            "Racing",
          ]
        }
        type="Cake"
      />
    </div>
    <div
      className="product__image-container"
    >
      <S3Image
        imgKey="/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584357819265-racecarcrop.jpg"
        theme={
          Object {
            "photoImg": Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
              "width": "100%",
            },
          }
        }
      />
    </div>
    <div
      className="new-product__button-container"
    >
      <Blueprint3.Button
        intent="danger"
        onClick={[Function]}
        style={
          Object {
            "margin": "8px 4px 0",
          }
        }
        text="Delete"
      />
      <Blueprint3.Button
        intent="warning"
        onClick={[Function]}
        style={
          Object {
            "margin": "8px 4px 0",
          }
        }
        text="Edit"
      />
    </div>
  </Blueprint3.Card>
  <Blueprint3.Alert
    canEscapeKeyCancel={false}
    canOutsideClickCancel={false}
    cancelButtonText="Cancel"
    confirmButtonText="Delete"
    icon="trash"
    intent="danger"
    isOpen={false}
    onCancel={[Function]}
    onConfirm={[Function]}
  >
    <p>
      Are you sure you want to delete "
      Racing Cake
      "?
    </p>
    <p>
      This cannot be undone.
    </p>
  </Blueprint3.Alert>
</Fragment>
`;

exports[`<ProductCard /> should render correctly when user IS admin and the product is of type "Creates" 1`] = `
<Fragment>
  <Blueprint3.Card
    className="product__card animate fadeIn"
    elevation={2}
    interactive={true}
    onClick={[Function]}
  >
    <i
      className="fas fa-pencil-alt product__icon"
      onClick={[Function]}
      role="button"
      style={
        Object {
          "color": "#9370f6",
        }
      }
      tabIndex={0}
    />
    <div
      className="product__text-container"
    >
      <div
        className="product__title"
      >
        <strong>
          Name Frame
        </strong>
      </div>
      <p
        className="product__price"
      >
        £14.50 + £4.50 postage
      </p>
      <TagsInput
        tags={
          Array [
            "Personalised",
            "Name",
            "Frame",
          ]
        }
        type="Creates"
      />
    </div>
    <div
      className="product__image-container"
    >
      <S3Image
        imgKey="/public/eu-west-2:1f7a6231-22c8-459a-a078-69ef91991a5e/1584106241635-IMG_0299.JPEG"
        theme={
          Object {
            "photoImg": Object {
              "alignItems": "center",
              "display": "flex",
              "justifyContent": "center",
              "width": "100%",
            },
          }
        }
      />
    </div>
    <div
      className="new-product__button-container"
    >
      <Blueprint3.Button
        intent="danger"
        onClick={[Function]}
        style={
          Object {
            "margin": "8px 4px 0",
          }
        }
        text="Delete"
      />
      <Blueprint3.Button
        intent="warning"
        onClick={[Function]}
        style={
          Object {
            "margin": "8px 4px 0",
          }
        }
        text="Edit"
      />
    </div>
  </Blueprint3.Card>
  <Blueprint3.Alert
    canEscapeKeyCancel={false}
    canOutsideClickCancel={false}
    cancelButtonText="Cancel"
    confirmButtonText="Delete"
    icon="trash"
    intent="danger"
    isOpen={false}
    onCancel={[Function]}
    onConfirm={[Function]}
  >
    <p>
      Are you sure you want to delete "
      Name Frame
      "?
    </p>
    <p>
      This cannot be undone.
    </p>
  </Blueprint3.Alert>
</Fragment>
`;
